src/Microsoft.ML.OnnxConverter/SaveOnnxCommand.cs

```diff
@@ -276,6 +303,11 @@ private void Run(IChannel ch)
                     Host.Assert(scorePipe.Source == end);
                     end = scorePipe;
                     transforms.AddLast(scoreOnnx);
+
+                    var xf = new KeyToValueMappingTransformer(Host, "PredictedLabel").Transform(scorePipe);
+                    var output = new CommonOutputs.TransformOutput { Model = new TransformModelImpl(Host, xf, scorePipe), OutputData = xf };
+                    end = output.OutputData;
+                    transforms.AddLast(end as ITransformCanSaveOnnx);
```

Comment:
So far, ignore everything in this PR except for this lines where I add the KeyToValueMapping at the end of the transforms list.

Here I had hardcoded the name of the PredictedLabel column, but I guess that simply retrieving the KeyDataViewType columns of the scorePipe dataview could be possible, so to then add KeyToValueMappingTransformers for each KeyDataViewType column we might desire.

Reply:
Ok, reading it again now. I like this idea because I think this code path runs only in the Nimbus scenario. (We should verify that, is is the case). If that is true, then this may unblock us for now. But I would like to be able to solve this for the non-Nimbus scenario as well. If this works for now, we can create a bug for the non-Nimbus scenario and fix it later. 

---
In reply to: [379220940](https://github.com/dotnet/machinelearning/pull/4841#discussion_r379220940) [](ancestors = 379220940)

Reply:
Yes, my understanding is that the Run() method where I added my changes only gets executed when running the following entrypoint:
[https://github.com/dotnet/machinelearning/blob/0b4b15bf551240f14b90d717b9b5be274caa526d/src/Microsoft.ML.OnnxConverter/SaveOnnxCommand.cs#L324-L329](https://github.com/dotnet/machinelearning/blob/0b4b15bf551240f14b90d717b9b5be274caa526d/src/Microsoft.ML.OnnxConverter/SaveOnnxCommand.cs#L324-L329)

So any tool that has access to that entrypoint will run that. I am unsure if NimbusML is the only one who has it (I wonder if MAML, and AutoML/ModelBuilder have access to it). If other tools have access to this, then perhaps adding an overload to Run() with a flag such as "addMappingAtTheEnd = false" might be a good idea.

On the other hand I am not sure how to confirm that this Run() method gets only called through that entrypoint. I might want to ask you about this offline about how to do it.



---
In reply to: [379256936](https://github.com/dotnet/machinelearning/pull/4841#discussion_r379256936) [](ancestors = 379256936,379220940)

Reply:
The command is runnable from MAML.

I added a comment in the [NimbusML issue](https://github.com/microsoft/NimbusML/issues/431), I am not sure that we should be adding transforms to the ONNX export, that do not exist in the model. If the goal is consistency between the results when scoring in NimbusML or in ONNX, to me it makes more sense to change NimbusML to make the `KeyToValue` transform a part of the model (otherwise, you will get consistency between NimbusML and ONNX, but not with ML.NET...).


---
In reply to: [379566192](https://github.com/dotnet/machinelearning/pull/4841#discussion_r379566192) [](ancestors = 379566192,379256936,379220940)

Reply:
I did try to add the KeyToValue transform to the pipeline from NimbusML. The problem is that the way NimbusML creates pipelines and run predictions in them, doesn't allow me to add a transform after the predictor. I tried a couple of approaches to make it work, but I couldn't manage to do it only in NimusML side. My conclusion of those attempts was that to add the KeyToValue transform from NimbusML side would also require to change the way that ML.NET executes graphs created in NimbusML (for instance the 

I believe that @ganik agreed that trying to change it from NimbusML (with the changes that would be required in ML.NET) would require more changes than simply do what I am doing here.

If you have something in mind to change it in NimbusML side without requiring major changes, please let me know. Thanks!

---
In reply to: [381234684](https://github.com/dotnet/machinelearning/pull/4841#discussion_r381234684) [](ancestors = 381234684,379566192,379256936,379220940)